// eslint-plugin-n8n-nodes-base/dist/lib/rules/node-param-type-options-missing-from-password.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var node_param_type_options_missing_from_password_exports = {};
__export(node_param_type_options_missing_from_password_exports, {
  default: () => node_param_type_options_missing_from_password_default
});
module.exports = __toCommonJS(node_param_type_options_missing_from_password_exports);
var import_utils = require("../ast/utils");
var import_getters = require("../ast/getters");
var node_param_type_options_missing_from_password_default = import_utils.utils.createRule({
  name: import_utils.utils.getRuleName(module),
  meta: {
    type: "problem",
    docs: {
      description: "`typeOptions.password` must be set to `true` in a Password node parameter, to obscure the password input.",
      recommended: "error"
    },
    fixable: "code",
    schema: [],
    messages: {
      addTypeOptionsPassword: "Add `typeOptions.password` [autofixable]"
    }
  },
  defaultOptions: [],
  create(context) {
    return {
      ObjectExpression(node) {
        const name = import_getters.getters.nodeParam.getName(node);
        if (!name || name.value !== "password")
          return;
        const typeOptions = import_getters.getters.nodeParam.getTypeOptions(node);
        if ((typeOptions == null ? void 0 : typeOptions.value.password) === true)
          return;
        const type = import_getters.getters.nodeParam.getType(node);
        if (!type)
          return;
        const { indentation, range } = import_utils.utils.getInsertionArgs(type);
        context.report({
          messageId: "addTypeOptionsPassword",
          node: type.ast,
          fix: (fixer) => fixer.insertTextAfterRange(
            range,
            `
${indentation}typeOptions: { password: true },`
          )
        });
      }
    };
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
